lib/plugins/output-filter/kubernetes-enrichment.js

```diff
@@ -0,0 +1,212 @@
+const { KubeConfig } = require('kubernetes-client')
+const Client = require('kubernetes-client').Client
+const Request = require('kubernetes-client/backends/request')
+const kubeconfig = new KubeConfig()
+var consoleLogger = require('../../util/logger.js')
+var podCache = {}
+var digestRegEx = /sha256:.+/i
+var client = null
+const FALSE_REGEX = /false/i
+if (process.env.KUBERNETES_PORT_443_TCP !== undefined) {
+  kubeconfig.loadFromCluster()
+} else {
+  kubeconfig.loadFromDefault()
+}
+client = new Client({ backend: new Request({ kubeconfig }) })
+client.loadSpec().then(() => {}).catch(error => {
+  console.error('Error in k8s client', error)
+})
+
+function kubernetesOutputFilter (context, config, eventEmitter, data, callback) {
+  // we use the config object to track state of each plugin instance
+  if (!config.client) {
+    initKubernetesClient(config)
+  } else {
+    enrichLogs(context, config, eventEmitter, data, callback)
+  }
+}
+
+function initKubernetesClient (config) {
+  // do we run in k8s cluster?
+  config.client = client
+  config.getPodSpec = function (namespace, podName, cb) {
+    this.client.api.v1.namespaces(namespace).pods(podName).get(cb)
+  }.bind(config)
+  // clean cache after 10 minutes
+  // TODO: use LRU Cache
+  setInterval(() => { podCache = {} }, 10 * 60 * 1000)
+}
+
+async function getPodSpec (config, namespace, podName, cb) {
```

Comment:
Why did you choose to use a callback instead of a return value? If you return a value, you can then use it while calling the next function, instead of complicating things with callbacks. Using `async-await` is meant to remove all callbacks from code.

Reply:
Yeah, next time ... Probably I'm too much used to 'old-style'. And `enrichLogs` would need to be declared as `async` and so on ... but the LA plugin framework does not expect a function declared as `async` for output-filters. Plus changes in error handling, which is also tricky. https://itnext.io/error-handling-with-async-await-in-js-26c3f20bc06a   
So in short I had to use `async/await` under the hood as the new Kubernetes API started to use `aync/await`, while in the general LA framework does not. Feel free to make improvements in a different PR.

Reply:
Aha, okay gotcha. Because this is open source and we want the community involved, writing in the latest, newest version of JavScript will attract contributions. But, also make us look good because the code is clean, with the latest features and ideally also trying to use functional programming and as much immutability as possible. All things that make the code more robust and we look better in the eyes of the community. Brand building etc...

Reply:
...and then at the same time you have actual businesses use Logagent in serious production environments and have to be careful with performance, stability, backwards compatibility, ensure things run not just on the latest and greatest VM, etc.